File: validator-b/spd_validator_b/policy.py
"""Shared contract data, not shared validator implementation."""
import json
import re
from functools import lru_cache
from pathlib import Path


@lru_cache(maxsize=1)
def policy():
    path = Path(__file__).resolve().parents[2] / 'validation/REQUIREMENT_ATTRIBUTION_0.1.2.json'
    if not path.is_file():
        path = Path(__file__).resolve().parent / 'data/REQUIREMENT_ATTRIBUTION_0.1.2.json'
    return json.loads(path.read_text(encoding='utf-8'))


def external_ids(output: str) -> list[str]:
    found = set()
    for line in output.splitlines():
        diagnostic = re.search(r'\b(?:ERROR|FATAL)\(([A-Z]+-\d+[a-z]?)\)', line)
        if diagnostic:
            found.add('SPD-BASE-002')
            for rule in policy()['externalDiagnostics']:
                if rule['code'] == diagnostic[1] and re.search(rule['pattern'], line):
                    found.update(rule['ids'])
    return sorted(found)

File: validator-a/src/policy.rs
//! Shared Contract 0.1.1 data; native validation is implemented in Rust.
use serde_json::Value;
use std::{collections::BTreeSet, sync::OnceLock};

pub fn data() -> &'static Value {
    static POLICY: OnceLock<Value> = OnceLock::new();
    POLICY.get_or_init(|| {
        serde_json::from_str(include_str!(
            "../../validation/REQUIREMENT_ATTRIBUTION_0.1.2.json"
        ))
        .expect("versioned attribution data")
    })
}

pub fn blocked(group: &str) -> Vec<String> {
    data()["blocked"][group]
        .as_array()
        .expect("known evidence group")
        .iter()
        .map(|v| v.as_str().unwrap().to_owned())
        .collect()
}

pub fn attribute(output: &str) -> Vec<String> {
    let diagnostic = regex::Regex::new(r"\b(?:ERROR|FATAL)\(([A-Z]+-\d+[a-z]?)\)").unwrap();
    let mut result = BTreeSet::new();
    for line in output.lines() {
        if let Some(found) = diagnostic.captures(line) {
            result.insert("SPD-BASE-002".to_owned());
            for rule in data()["externalDiagnostics"].as_array().unwrap() {
                if rule["code"].as_str() == Some(&found[1])
                    && regex::Regex::new(rule["pattern"].as_str().unwrap())
                        .unwrap()
                        .is_match(line)
                {
                    for id in rule["ids"].as_array().unwrap() {
                        result.insert(id.as_str().unwrap().to_owned());
                    }
                }
            }
        }
    }
    result.into_iter().collect()
}
